core/src/block_creation_loop.rs

```diff
@@ -448,43 +583,81 @@ fn shutdown_and_drain_record_receiver(
 /// - Insert the alpentick
 /// - Clear the working bank
 fn record_and_complete_block(
-    poh_recorder: &RwLock<PohRecorder>,
-    record_receiver: &mut RecordReceiver,
-    block_timer: Instant,
+    ctx: &mut LeaderContext,
+    mut optimistic_parent: Option<(Slot, Hash)>,
+    block_timer: &mut Option<Instant>,
     block_timeout: Duration,
 ) -> Result<(), PohRecorderError> {
     loop {
-        let remaining_slot_time = block_timeout.saturating_sub(block_timer.elapsed());
-        if remaining_slot_time.is_zero() {
+        // Check if we've exceeded the block timeout
+        let remaining = time_left(*block_timer, block_timeout);
+        if remaining.is_some_and(|t| t.is_zero()) {
             break;
         }
-
-        let Ok(record) = record_receiver.recv_timeout(remaining_slot_time) else {
-            continue;
+        let timeout = remaining.unwrap_or(Duration::from_secs(1));
+
+        // Wait for either a record or (if optimistic) parent ready notification
+        let result = if optimistic_parent.is_some() {
+            select_biased! {
+                recv(ctx.leader_window_info_receiver) -> msg => {
+                    msg.ok().map(RecvResult::ParentReady).unwrap_or(RecvResult::Timeout)
+                },
+                recv(ctx.record_receiver.inner()) -> msg => {
+                    if let Ok(record) = msg {
+                        ctx.record_receiver
+                            .on_received_record(record.transaction_batches.len() as u64);
+                        RecvResult::Record(record)
+                    } else {
+                        RecvResult::Timeout
+                    }
+                },
+                default(timeout) => RecvResult::Timeout,
+            }
+        } else {
+            ctx.record_receiver
+                .recv_timeout(timeout)
+                .map(RecvResult::Record)
+                .unwrap_or(RecvResult::Timeout)
         };
-        poh_recorder.write().unwrap().record(
-            record.slot,
-            record.mixins,
-            record.transaction_batches,
-        )?;
+
+        match result {
+            RecvResult::Record(record) => {
+                ctx.poh_recorder.write().unwrap().record(
+                    record.slot,
+                    record.mixins,
+                    record.transaction_batches,
+                )?;
+            }
+            RecvResult::ParentReady(leader_window_info) => {
+                // UNWRAP: the select_biased! above only produces ParentReady if optimistic_parent
```

Comment:
i think with my suggestion above we also combine this into the select (remove the intermediate `result`) and then wed on't have to worry about this